src/main/java/net/minecraftforge/registries/RegistryObject.java

```diff
@@ -297,32 +303,68 @@ public void accept(Predicate<ResourceLocation> pred)
     }
 
     /**
-     * Directly retrieves the wrapped Registry Object. This value will automatically be updated when the backing registry is updated.
-     * Will throw NPE if the value is null, use isPresent to check first. Or use any of the other guarded functions.
+     * Retrieves the wrapped object in the registry.
+     * This value will automatically be updated when the backing registry is updated.
+     *
+     * @throws NullPointerException If the value is null. Use {@link #isPresent()} to check if the value exists first.
+     * @see #isPresent()
+     * @see #orElse(Object)
+     * @see #orElseGet(Supplier)
+     * @see #orElseThrow(Supplier)
      */
+    @NotNull
     @Override
-    @Nonnull
     public T get()
     {
         T ret = this.value;
         Objects.requireNonNull(ret, () -> "Registry Object not present: " + this.name);
         return ret;
     }
 
+    @SuppressWarnings("unchecked")
     @Deprecated(since = "1.18.1") // TODO: make package-private
     public void updateReference(IForgeRegistry<? extends T> registry)
     {
-        this.value = registry.containsKey(this.name) ? registry.getValue(this.name) : null;
+        if (this.name == null)
+            return;
+        if (registry.containsKey(this.name))
+        {
+            this.value = registry.getValue(this.name);
+            if (this.key == null)
+                this.key = (ResourceKey<T>) ResourceKey.create(registry.getRegistryKey(), this.name);
+            this.holder = (Holder<T>) registry.getHolder(this.name).orElse(null);
+        }
+        else
+        {
+            this.value = null;
+            this.holder = null;
+        }
     }
 
+    @SuppressWarnings("unchecked")
     void updateReference(Registry<? extends T> registry)
     {
-        this.value = registry.containsKey(this.name) ? registry.get(this.name) : null;
+        if (this.name == null)
+            return;
+        if (registry.containsKey(this.name))
+        {
+            this.value = registry.get(this.name);
+            if (this.key == null)
+                this.key = (ResourceKey<T>) ResourceKey.create(registry.key(), this.name);
+            this.holder = ((Registry<T>) registry).getHolder(ResourceKey.create((ResourceKey<? extends Registry<T>>) registry.key(), this.name)).orElse(null);
```

Comment:
Is there a reason this isn't just:
```java
this.holder = ((Registry<T>) registry).getHolder(this.key).orElse(null)
```
as if the key was null, then it is initialized to the same value the line above, and otherwise we already have it?

Reply:
No reason probably just missed this when I was modifying stuff